### ISSUE TITLE
Creating the MR on the forked GitLab project if the project is forked using `target_project_id`

### ISSUE THREAD

**Issue by valdisrigdon** [renovatebot/renovate#9107]
**What would you like Renovate to be able to do?**

Attempting to run with `includeForks: true` so that a forked repository contains the renovate branches. Our setup includes the upstream repository having protected branches and supports a development workflow where all developer branches are created on forked repositories but PRs are targeted to the upstream repository.

**Did you already have any implementation ideas?**

The [GitLab API for creating a MR](https://docs.gitlab.com/ee/api/merge_requests.html#create-mr) supports `target_project_id`. The upstream repository id and then passed along to the create MR API call.

**Comment by rarkins** [comment]
Please describe the workflow in detail.

For example, what reasons are there for not running Renovate against the parent repo to begin with? Did you create the fork just for the reason of Renovate, or it exists for some other reason?

We have the concept of `forkMode` for GitHub, where it *creates* the fork if necessary under its own account.

**Comment by valdisrigdon** [comment]
I've updated the description with a few more details!

**Comment by rarkins** [comment]
> I don't think this requires any configuration, but rather if `includeForks == true` then the MR should really be targeted against the forked repository.

No, because not everybody has the same workflow as you, and it would be an unwelcome change of behavior for many. It definitely would require configuration if implemented.

There are a lot of *other* platform interactions other than just "create MR" that need to be considered. The full list is [here](https://github.com/renovatebot/renovate/blob/master/lib/platform/types.ts#L138-L178) and someone would need to go through each one to work out if it should be run against the parent or fork.

When you want Renovate to run on an existing fork that you or someone else has already created, who or what is responsible for keeping the forked repo base branch up to date with the parent?

**Comment by valdisrigdon** [comment]
> > I don't think this requires any configuration, but rather if `includeForks == true` then the MR should really be targeted against the forked repository.
> 
> No, because not everybody has the same workflow as you, and it would be an unwelcome change of behavior for many. It definitely would require configuration if implemented.
> 

That's a great point, yes, a config option would be required.

> There are a lot of _other_ platform interactions other than just "create MR" that need to be considered. The full list is [here](https://github.com/renovatebot/renovate/blob/master/lib/platform/types.ts#L138-L178) and someone would need to go through each one to work out if it should be run against the parent or fork.

That list is quite long. The MR exists on the upstream fork, so all "MR" actions would expect the MR to exist there. We don't use GitLab issues, so I'm unsure about that. The branch exists on the forked repository, so "branch" actions would go against the source repository.

> When you want Renovate to run on an existing fork that you or someone else has already created, who or what is responsible for keeping the forked repo base branch up to date with the parent?

The fork is already created and the default branch is pull mirrored from the upstream repository. As I write that, I realize that by default renovate is skipping mirrored repositories -- is that configurable?

As I write this reply, it seems that this is a much larger request than I originally thought it was.

Another approach that would work for us would be to push branches to the upstream repository, but since all branches are protected, we would have to use a Deploy Token that's allowed to push to protected branches, which GitLab has enabled once again recently.  However, since renovate doesn't support the `ssh` protocol with GitLab, the option to have renovate create branches in our upstream repository failed as well.

**Comment by rarkins** [comment]
> The fork is already created and the default branch is pull mirrored from the upstream repository. As I write that, I realize that by default renovate is skipping mirrored repositories -- is that configurable?

It's not, but it probably could be. What is the relationship between forked and mirrored repositories though- can a repository be *both* in GitLab? I had expected it to be either, not both.

> However, since renovate doesn't support the `ssh` protocol with GitLab, the option to have renovate create branches in our upstream repository failed as well.

I don't anticipate that would be a huge thing to implement, once requirements are settled. In that case though you'd run it on the parent project and not on a fork/mirror?

**Comment by valdisrigdon** [comment]
> > The fork is already created and the default branch is pull mirrored from the upstream repository. As I write that, I realize that by default renovate is skipping mirrored repositories -- is that configurable?
> 
> It's not, but it probably could be. What is the relationship between forked and mirrored repositories though- can a repository be _both_ in GitLab? I had expected it to be either, not both.

There are two repositories and both are in GitLab, in different GitLab groups. `upstream` contains only the default branch, and all branches are protected. All development is done on the `dev` fork, which contains all the feature branches for all developers, forked from `upstream`, and protected branches from `upstream` are mirrored to `dev`. When an MR is opened from the GitLab UI, they default to targeting `upstream/master`.

> 
> > However, since renovate doesn't support the `ssh` protocol with GitLab, the option to have renovate create branches in our upstream repository failed as well.
> 
> I don't anticipate that would be a huge thing to implement, once requirements are settled. In that case though you'd run it on the parent project and not on a fork/mirror?

Yes, if we could support the `ssh` protocol, then then I believe we could run renovate on the `upstream` repository as the Deploy Token could push to protected branches.  It's not ideal, but it would get us started. It would be preferable to run on the forked repo, but as noted above, there are a lot of changes required for that.

**Comment by rarkins** [comment]
To confirm I understand: developers all share and have write access to the dev project? I.e. you don't all have your own forks but instead a shared fork? And all MRs are raised directly against upstream instead of merged to dev first?

**Comment by valdisrigdon** [comment]
> To confirm I understand: developers all share and have write access to the dev project? I.e. you don't all have your own forks but instead a shared fork? And all MRs are raised directly against upstream instead of merged to dev first?

This is correct.
